stackslib/src/net/p2p.rs

```diff
@@ -4295,38 +4295,55 @@ impl PeerNetwork {
         }
     }
 
-    /// Refresh our view of the last three reward cycles
-    /// This ensures that the PeerNetwork has cached copies of the reward cycle data (including the
-    /// signing set) for the current, previous, and previous-previous reward cycles.  This data is
-    /// in turn consumed by the Nakamoto block downloader, which must validate blocks signed from
-    /// any of these reward cycles.
-    #[cfg_attr(test, mutants::skip)]
-    fn refresh_reward_cycles(
-        &mut self,
+    /// Determine if we need to invalidate a given cached reward set.
+    ///
+    /// In Epoch 2, this requires checking the first sortition in the start of the reward set's
+    /// reward phase.
+    ///
+    /// In Nakamoto, this requires checking the anchor block in the prepare phase for the upcoming
+    /// reward phase.
+    fn check_reload_cached_reward_set(
+        &self,
         sortdb: &SortitionDB,
-        chainstate: &mut StacksChainState,
+        chainstate: &StacksChainState,
+        rc: u64,
         tip_sn: &BlockSnapshot,
         tip_block_id: &StacksBlockId,
-    ) -> Result<(), net_error> {
-        let cur_rc = self
-            .burnchain
-            .block_height_to_reward_cycle(tip_sn.block_height)
-            .expect("FATAL: sortition from before system start");
-
-        let prev_rc = cur_rc.saturating_sub(1);
-        let prev_prev_rc = prev_rc.saturating_sub(1);
-        let ih = sortdb.index_handle(&tip_sn.sortition_id);
-
-        for rc in [cur_rc, prev_rc, prev_prev_rc] {
-            debug!("Refresh reward cycle info for cycle {}", rc);
+        tip_height: u64,
+    ) -> Result<bool, net_error> {
+        let epoch = self.get_epoch_at_burn_height(tip_sn.block_height);
+        if epoch.epoch_id >= StacksEpochId::Epoch30 {
+            // epoch 3, where there are no forks except from bugs or burnchain reorgs.
+            // invalidate reward cycles on burnchain or stacks reorg, should they ever happen
+            let reorg = Self::is_reorg(Some(&self.burnchain_tip), tip_sn, sortdb)
+                || Self::is_nakamoto_reorg(
+                    &self.stacks_tip.block_id(),
+                    self.stacks_tip.height,
+                    tip_block_id,
+                    tip_height,
+                    chainstate,
+                );
+            if reorg {
+                info!("Burnchain or Stacks reorg detected; will invalidate cached reward set for cycle {rc}");
+            }
```

Comment:
This info line will alway follow another info line during the reorg check, right? One of them should be dropped.

Reply:
Fixed